File: python/src/linodemcp/profiles/scopecheck.py
"""Scope comparison primitives for Phase 6.4 token validation.

Mirrors ``go/internal/profiles/scopecheck.go``. Pure functions: parse a
PAT scope string into a scope set, flatten an OAuth Grants payload into
the same set shape, and compare a profile's required scopes against
what the token carries.

The pipeline compares plain scope strings, matching Go's open Scope
type: a token or a profile may name scopes the catalog does not spell
(child_account:*, the declared per-tool families, future Linode
scopes), and the comparison matches them by value. The Scope enum only
names the values the grants converter produces.

The loader (wired in Phase 6.4b) takes a comparison result and decides
policy: missing scopes are a hard failure, excess scopes warn unless
strict mode promotes them to errors.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from linodemcp.profiles.scope import Scope

if TYPE_CHECKING:
    from collections.abc import Sequence

    from linodemcp.linode import GlobalGrants, Grant, Grants


def parse_pat_scopes(scope_str: str) -> list[str]:
    """Split a Linode PAT scope string into a deduplicated sorted list.

    The format is space-delimited tokens; ``"*"`` stands for "all
    permissions" on every category. Whitespace-only or empty input
    yields an empty list. Every token is kept as-is, catalog member or
    not: a PAT carries whatever scopes Linode granted it, and dropping
    one here would report the token as under-scoped for a value it
    genuinely holds.
    """
    return sorted(set(scope_str.split()))


def _add_pair(perm: str, read_only: Scope, read_write: Scope, seen: set[Scope]) -> None:
    """Add the scope pair implied by a permission string.

    Mirrors Go's addPair helper: ``"read_write"`` implies both scopes;
    ``"read_only"`` implies only the read-only scope; anything else
    (empty string, unknown value) contributes nothing.
    """
    if perm == "read_write":
        seen.add(read_write)
        seen.add(read_only)
    elif perm == "read_only":
        seen.add(read_only)


def _collect_global(global_: GlobalGrants, seen: set[Scope]) -> None:
    """Walk the OAuth global grant booleans and add the implied scopes.

    AccountAccess gives the account read_only/read_write pair directly;
    each ``add_<resource>`` boolean implies the corresponding category's
    :read_write (write subsumes read).
    """
    _add_pair(
        global_.account_access,
        Scope.AccountReadOnly,
        Scope.AccountReadWrite,
        seen,
    )

    pairs = [
        (global_.add_linodes, Scope.LinodesReadOnly, Scope.LinodesReadWrite),
        (global_.add_domains, Scope.DomainsReadOnly, Scope.DomainsReadWrite),
        (global_.add_firewalls, Scope.FirewallReadOnly, Scope.FirewallReadWrite),
        (global_.add_images, Scope.ImagesReadOnly, Scope.ImagesReadWrite),
        (global_.add_databases, Scope.DatabasesReadOnly, Scope.DatabasesReadWrite),
        (global_.add_longview, Scope.LongviewReadOnly, Scope.LongviewReadWrite),
        (
            global_.add_nodebalancers,
            Scope.NodeBalancersReadOnly,
            Scope.NodeBalancersReadWrite,
        ),
        (
            global_.add_stackscripts,
            Scope.StackScriptsReadOnly,
            Scope.StackScriptsReadWrite,
        ),
        (global_.add_volumes, Scope.VolumesReadOnly, Scope.VolumesReadWrite),
        (global_.add_vpcs, Scope.VPCReadOnly, Scope.VPCReadWrite),
    ]
    for has_flag, read_only, read_write in pairs:
        if has_flag:
            seen.add(read_write)
            seen.add(read_only)


def _collect_resources(grants: Grants, seen: set[Scope]) -> None:
    """Walk the per-resource grant lists and add the scopes they imply."""
    category_map: list[tuple[list[Grant], Scope, Scope]] = [
        (list(grants.linode), Scope.LinodesReadOnly, Scope.LinodesReadWrite),
        (list(grants.domain), Scope.DomainsReadOnly, Scope.DomainsReadWrite),
        (
            list(grants.nodebalancer),
            Scope.NodeBalancersReadOnly,
            Scope.NodeBalancersReadWrite,
        ),
        (list(grants.image), Scope.ImagesReadOnly, Scope.ImagesReadWrite),
        (list(grants.longview), Scope.LongviewReadOnly, Scope.LongviewReadWrite),
        (
            list(grants.stackscript),
            Scope.StackScriptsReadOnly,
            Scope.StackScriptsReadWrite,
        ),
        (list(grants.volume), Scope.VolumesReadOnly, Scope.VolumesReadWrite),
        (list(grants.database), Scope.DatabasesReadOnly, Scope.DatabasesReadWrite),
        (list(grants.firewall), Scope.FirewallReadOnly, Scope.FirewallReadWrite),
        (list(grants.vpc), Scope.VPCReadOnly, Scope.VPCReadWrite),
        (list(grants.lkecluster), Scope.LKEReadOnly, Scope.LKEReadWrite),
    ]
    for entries, read_only, read_write in category_map:
        for entry in entries:
            _add_pair(entry.permissions, read_only, read_write, seen)


def flatten_grants(grants: Grants | None) -> list[Scope]:
    """Walk a /profile/grants response and return the effective Scope set.

    Global account booleans produce account-level scopes; per-resource
    grants produce the matching category scope. Empty/null permission
    entries contribute nothing. The output is sorted and deduplicated.
    A ``None`` input yields an empty list (matches Go's nil-safe path).
    """
    if grants is None:
        return []

    seen: set[Scope] = set()
    _collect_global(grants.global_, seen)
    _collect_resources(grants, seen)

    return sorted(seen, key=lambda s: s.value)


@dataclass(frozen=True)
class ScopeComparison:
    """Result of comparing a profile's required scopes against actuals.

    ``missing`` lists scopes the profile requires but the token lacks;
    these are a hard failure at load time. ``excess`` lists scopes the
    token carries that the profile doesn't require; warning by default,
    strict mode promotes them to errors.
    """

    missing: tuple[str, ...] = field(default_factory=tuple)
    excess: tuple[str, ...] = field(default_factory=tuple)

    @property
    def has_missing(self) -> bool:
        """True if the token is under-scoped for the active profile."""
        return bool(self.missing)

    @property
    def has_excess(self) -> bool:
        """True if the token carries more access than the profile needs."""
        return bool(self.excess)


def compare_scopes(required: Sequence[str], actual: Sequence[str]) -> ScopeComparison:
    """Return the missing and excess sets between required and actual.

    Set-based; order doesn't matter, and both sides are plain scope
    strings so values outside the catalog compare like any other.
    Output is sorted ascending so error messages stay stable. Wildcard
    handling:

    - ``Scope.Wildcard`` ("*") in ``actual`` matches every required
      scope. A token with just ``"*"`` satisfies any profile.
    - ``Scope.Wildcard`` in ``required`` is ignored (not meaningful for
      derived profiles; user-defined profiles may include it).
    """
    actual_set = set(actual)
    has_wildcard = Scope.Wildcard in actual_set

    required_set = {s for s in required if s != Scope.Wildcard}

    missing: list[str] = []
    if not has_wildcard:
        missing = sorted(s for s in required_set if s not in actual_set)

    excess = sorted(
        s for s in actual_set if s != Scope.Wildcard and s not in required_set
    )

    return ScopeComparison(missing=tuple(missing), excess=tuple(excess))

File: go/internal/profiles/scopecheck.go
package profiles

import (
	"slices"
	"strings"

	"github.com/chadit/LinodeMCP/internal/linode"
)

// ParsePATScopes splits a Linode PAT scope string (the value returned in
// the /profile response's `scopes` field) into a deduplicated Scope set.
// The format is space-delimited tokens; "*" stands for "all permissions"
// on the matching category. Empty/whitespace input yields an empty set.
func ParsePATScopes(scopeStr string) []Scope {
	fields := strings.Fields(scopeStr)
	if len(fields) == 0 {
		return nil
	}

	seen := make(map[Scope]struct{}, len(fields))
	for _, f := range fields {
		seen[Scope(f)] = struct{}{}
	}

	out := make([]Scope, 0, len(seen))
	for s := range seen {
		out = append(out, s)
	}

	slices.Sort(out)

	return out
}

// FlattenGrants walks a /profile/grants response and returns the
// effective set of Scope values it grants. Global account booleans
// produce account-level scopes; per-resource grants imply the matching
// category scope. The returned slice is sorted and deduplicated.
//
// The mapping is conservative: a grant present at any permission level
// implies the matching :read_only scope; permission "read_write" also
// implies :read_write. Empty/null permission entries grant nothing.
func FlattenGrants(grants *linode.Grants) []Scope {
	if grants == nil {
		return nil
	}

	seen := make(map[Scope]struct{})
	collectGlobal(grants.Global, seen)
	collectResourceGrants(grants, seen)

	out := make([]Scope, 0, len(seen))
	for s := range seen {
		out = append(out, s)
	}

	slices.Sort(out)

	return out
}

// collectGlobal adds account-level scopes for each true bool on
// GlobalGrants. AccountAccess gives the :read_only/:read_write pair
// directly; the add_* flags imply the resource category's :read_write
// (a token that can "add linodes" can write linodes).
func collectGlobal(global linode.GlobalGrants, seen map[Scope]struct{}) {
	switch global.AccountAccess {
	case "read_write":
		seen[ScopeAccountReadWrite] = struct{}{}
		seen[ScopeAccountReadOnly] = struct{}{}
	case "read_only":
		seen[ScopeAccountReadOnly] = struct{}{}
	}

	if global.AddLinodes {
		seen[ScopeLinodesReadWrite] = struct{}{}
		seen[ScopeLinodesReadOnly] = struct{}{}
	}

	if global.AddDomains {
		seen[ScopeDomainsReadWrite] = struct{}{}
		seen[ScopeDomainsReadOnly] = struct{}{}
	}

	if global.AddFirewalls {
		seen[ScopeFirewallReadWrite] = struct{}{}
		seen[ScopeFirewallReadOnly] = struct{}{}
	}

	if global.AddImages {
		seen[ScopeImagesReadWrite] = struct{}{}
		seen[ScopeImagesReadOnly] = struct{}{}
	}

	if global.AddDatabases {
		seen[ScopeDatabasesReadWrite] = struct{}{}
		seen[ScopeDatabasesReadOnly] = struct{}{}
	}

	if global.AddLongview {
		seen[ScopeLongviewReadWrite] = struct{}{}
		seen[ScopeLongviewReadOnly] = struct{}{}
	}

	if global.AddNodeBalancers {
		seen[ScopeNodeBalancersReadWrite] = struct{}{}
		seen[ScopeNodeBalancersReadOnly] = struct{}{}
	}

	if global.AddStackScripts {
		seen[ScopeStackScriptsReadWrite] = struct{}{}
		seen[ScopeStackScriptsReadOnly] = struct{}{}
	}

	if global.AddVolumes {
		seen[ScopeVolumesReadWrite] = struct{}{}
		seen[ScopeVolumesReadOnly] = struct{}{}
	}

	if global.AddVPCs {
		seen[ScopeVPCReadWrite] = struct{}{}
		seen[ScopeVPCReadOnly] = struct{}{}
	}
}

// collectResourceGrants walks the per-resource grant lists and adds the
// matching category scopes for any non-empty permission. A grant with
// permission "read_write" produces both read_only and read_write; a
// grant with "read_only" produces only :read_only.
func collectResourceGrants(grants *linode.Grants, seen map[Scope]struct{}) {
	addPair := func(perm linode.GrantPermission, readOnly, readWrite Scope) {
		switch perm {
		case "read_write":
			seen[readWrite] = struct{}{}
			seen[readOnly] = struct{}{}
		case "read_only":
			seen[readOnly] = struct{}{}
		}
	}

	for _, g := range grants.Linode {
		addPair(g.Permissions, ScopeLinodesReadOnly, ScopeLinodesReadWrite)
	}

	for _, g := range grants.Domain {
		addPair(g.Permissions, ScopeDomainsReadOnly, ScopeDomainsReadWrite)
	}

	for _, g := range grants.NodeBalancer {
		addPair(g.Permissions, ScopeNodeBalancersReadOnly, ScopeNodeBalancersReadWrite)
	}

	for _, g := range grants.Image {
		addPair(g.Permissions, ScopeImagesReadOnly, ScopeImagesReadWrite)
	}

	for _, g := range grants.Longview {
		addPair(g.Permissions, ScopeLongviewReadOnly, ScopeLongviewReadWrite)
	}

	for _, g := range grants.StackScript {
		addPair(g.Permissions, ScopeStackScriptsReadOnly, ScopeStackScriptsReadWrite)
	}

	for _, g := range grants.Volume {
		addPair(g.Permissions, ScopeVolumesReadOnly, ScopeVolumesReadWrite)
	}

	for _, g := range grants.Database {
		addPair(g.Permissions, ScopeDatabasesReadOnly, ScopeDatabasesReadWrite)
	}

	for _, g := range grants.Firewall {
		addPair(g.Permissions, ScopeFirewallReadOnly, ScopeFirewallReadWrite)
	}

	for _, g := range grants.VPC {
		addPair(g.Permissions, ScopeVPCReadOnly, ScopeVPCReadWrite)
	}

	for _, g := range grants.LKECluster {
		addPair(g.Permissions, ScopeLKEReadOnly, ScopeLKEReadWrite)
	}
}

// ScopeComparison captures the result of comparing a profile's required
// scope list against a token's actual scope list. Missing scopes are a
// hard failure (the token can't do what the profile needs). Excess
// scopes are a least-privilege warning by default; strict mode promotes
// them to errors.
type ScopeComparison struct {
	// Missing lists scopes the profile requires but the token lacks.
	// Sorted ascending so error messages stay stable.
	Missing []Scope

	// Excess lists scopes the token carries that the profile does not
	// require. Sorted ascending.
	Excess []Scope
}

// HasMissing reports whether the token is under-scoped for the active
// profile. Phase 6.4 treats this as a hard failure at load time.
func (c ScopeComparison) HasMissing() bool { return len(c.Missing) > 0 }

// HasExcess reports whether the token carries more access than the
// active profile asks for. Warning by default; strict mode treats it
// as a failure.
func (c ScopeComparison) HasExcess() bool { return len(c.Excess) > 0 }

// CompareScopes returns the missing/excess sets between a profile's
// required scopes and the token's actual scopes. Wildcard handling:
//   - ScopeWildcard ("*") in `actual` matches any required scope.
//   - A required scope of "*" is not meaningful (callers should keep
//     this out of profile definitions) but is treated as "no specific
//     scope required" so an empty required slice and a `["*"]` required
//     slice both pass.
//
// Comparison is set-based: order of inputs doesn't matter. Output is
// sorted ascending so error messages and test fixtures stay stable.
func CompareScopes(required, actual []Scope) ScopeComparison {
	actualSet := make(map[Scope]struct{}, len(actual))

	var hasWildcard bool

	for _, s := range actual {
		actualSet[s] = struct{}{}
		if s == ScopeWildcard {
			hasWildcard = true
		}
	}

	requiredSet := make(map[Scope]struct{}, len(required))
	for _, s := range required {
		if s == ScopeWildcard {
			continue
		}

		requiredSet[s] = struct{}{}
	}

	var missing []Scope

	if !hasWildcard {
		for scope := range requiredSet {
			if _, ok := actualSet[scope]; !ok {
				missing = append(missing, scope)
			}
		}
	}

	var excess []Scope

	for scope := range actualSet {
		if scope == ScopeWildcard {
			continue
		}

		if _, ok := requiredSet[scope]; !ok {
			excess = append(excess, scope)
		}
	}

	slices.Sort(missing)
	slices.Sort(excess)

	return ScopeComparison{Missing: missing, Excess: excess}
}
